### PR TITLE
Fix: set kryo options before loading registrars

### LINKED ISSUE #1204: KryoAtomicCoder overrides setRegistrationRequired
It looks like a change in a recent behavior unless we set a flag in pipeline options, 

```
opts.setKryoRegistrationRequired(true)
```

Example code:

```scala
 @KryoRegistrar
 class MyKryoRegistrar extends IKryoRegistrar {
   def apply(k: Kryo): Unit = {
     k.setRegistrationRequired(true)
   }
}
```

Expected to throw an exception, but doesn't:

```scala
case class Foo()
val coder = com.spotify.scio.util.ScioUtil.getScalaCoder[Foo]
CoderUtils.encodeToByteArray(coder, Foo())
```

### PR DESCRIPTION
**Pull request by regadas**
Fixes #1204

### CODE REVIEW

**Comment by kanterov** [comment]
Can you add a test that will catch this? Something similar to what was reported in the original issue.

**Comment by jto** [comment]
LGTM

File: scio-core/src/main/scala/com/spotify/scio/util/Functions.scala [hunk 1]
```diff
@@ -197,4 +199,12 @@ private[scio] object Functions {
     override def reduceOption(accumulator: Iterable[T]): Option[T] =
       _mon.sumOption(accumulator).orElse(Some(_mon.zero))
   }
+
+  def memoize[T, U](f: T => U): T => U = {
+    lazy val cache = new ConcurrentHashMap[T, U]()
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/util/Functions.scala]: How many `cache` instances will this produce? Will it become a member of `object Functions`?

File: scio-test/src/test/scala/com/spotify/scio/coders/KryoAtomicCoderTest.scala [hunk 2]
```diff
@@ -38,87 +38,85 @@ import scala.collection.JavaConverters._
 import scala.reflect.ClassTag
 
 class KryoAtomicCoderTest extends PipelineSpec {
-
```
Comment (nevillelyh) [inline scio-test/src/test/scala/com/spotify/scio/coders/KryoAtomicCoderTest.scala]: Can we undo all the unnecessary style change here?

Comment (regadas) [reply scio-test/src/test/scala/com/spotify/scio/coders/KryoAtomicCoderTest.scala]: yup

File: scio-core/src/main/scala/com/spotify/scio/util/Functions.scala (hunk 1 above)
Comment (regadas) [reply scio-core/src/main/scala/com/spotify/scio/util/Functions.scala]: this will produce one `cache` instance per function. I don't believe so.

**Pull request merged**